py_env_create and a test broken after Derecho updates to conda/mamba

### Brief summary of bug

It seems that recent conda and mamba updates on Derecho are now causing failures in two Python system tests exercising `py_env_create`.

For the future, it would be good for a few of us to have a nightly cron job that exercises any Python tests that rely on the Derecho environment. That would include at least these system tests:
- `test_sys_py_env_create.py` (conda and mamba)
- `test_sys_mesh_modifier.py` (ncks)

### General bug information

**CTSM version you are using:** ctsm5.3.067

**Does this bug cause significantly incorrect results in the model's science?** No

**Configurations affected:** n/a

### Details of bug

`test_complete_py_env_create` fails because `conda rename` doesn't accept the `--force` argument as of conda 25.5.51. This is a true bug. Looks like that functionality was [replaced](https://github.com/conda/conda/blob/408abc3a3826c80750883accb8896005e3ac97b2/CHANGELOG.md?plain=1#L291) with `--yes` in conda 25.3.0.

`test_complete_py_env_create_mamba` fails because `mamba env create` can't accept empty files via `--file` at mamba 2.3.1. Note that this is just a test failure, but it might be masking functionality failure.

These seem to have been working as of mamba 1.5.9 / conda 24.7.1.

### Important output or errors that show the problem

`test_complete_py_env_create`:
```
FAIL: test_complete_py_env_create (test.test_sys_py_env_create.TestSysPyEnvCreate.test_complete_py_env_create)
A few calls of py_env_create to ensure it's working right.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/glade/work/samrabin/ctsm/python/ctsm/test/test_sys_py_env_create.py", line 319, in test_complete_py_env_create
    _, out = self._create_empty_env(
             ~~~~~~~~~~~~~~~~~~~~~~^
        new_env_name=False, extra_args=["-r", self.env_names[-1] + "_orig"]
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/glade/work/samrabin/ctsm/python/ctsm/test/test_sys_py_env_create.py", line 109, in _create_empty_env
    self.assertEqual(
    ~~~~~~~~~~~~~~~~^
        out.returncode, 0, "Unexpected failure of py_env_create call:\n" + out.stderr
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
AssertionError: 2 != 0 : Unexpected failure of py_env_create call:
+ conda rename --force -n pec_test_BVhWy pec_test_BVhWy_orig
usage: conda [-h] [-v] [--no-plugins] [-V] COMMAND ...
conda: error: unrecognized arguments: --force


Stdout:
cmd: /glade/work/samrabin/ctsm/py_env_create -n pec_test_BVhWy -f /glade/work/samrabin/ctsm/python/empty.txt --yes
STDOUT:
Using conda to install the python environment needed to run the CTSM python tools in the conda environment: pec_test_BVhWy
Using the file: /glade/work/samrabin/ctsm/python/empty.txt
Installing pec_test_BVhWy; this can take a few minutes.
Channels:
 - conda-forge
 - r
Platform: linux-64
Collecting package metadata (repodata.json): done
Solving environment: done

Downloading and Extracting Packages: ...working... done
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate pec_test_BVhWy
#
# To deactivate an active environment, use
#
#     $ conda deactivate

Successfully installed the pec_test_BVhWy python environment

activate the environment by doing the following...
conda activate pec_test_BVhWy

STDERR:
+ conda env create --yes --file /glade/work/samrabin/ctsm/python/empty.txt --name pec_test_BVhWy

RETURN CODE: 0
cmd: /glade/work/samrabin/ctsm/py_env_create -n pec_test_BVhWy -f /glade/work/samrabin/ctsm/python/empty.txt --yes -o
STDOUT:
Using conda to install the python environment needed to run the CTSM python tools in the conda environment: pec_test_BVhWy
Using the file: /glade/work/samrabin/ctsm/python/empty.txt
Overwriting existing pec_test_BVhWy

Remove all packages in environment /glade/work/samrabin/conda-envs/pec_test_BVhWy:

No packages found in /glade/work/samrabin/conda-envs/pec_test_BVhWy. Continuing environment removal
Installing pec_test_BVhWy; this can take a few minutes.
Channels:
 - conda-forge
 - r
Platform: linux-64
Collecting package metadata (repodata.json): done
Solving environment: done

Downloading and Extracting Packages: ...working... done
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate pec_test_BVhWy
#
# To deactivate an active environment, use
#
#     $ conda deactivate

Successfully installed the pec_test_BVhWy python environment

activate the environment by doing the following...
conda activate pec_test_BVhWy

STDERR:
+ conda remove --yes -n pec_test_BVhWy --all
+ conda env create --yes --file /glade/work/samrabin/ctsm/python/empty.txt --name pec_test_BVhWy

RETURN CODE: 0
cmd: /glade/work/samrabin/ctsm/py_env_create -n pec_test_BVhWy -f /glade/work/samrabin/ctsm/python/empty.txt --yes -r pec_test_BVhWy_orig
STDOUT:
Using conda to install the python environment needed to run the CTSM python tools in the conda environment: pec_test_BVhWy
Using the file: /glade/work/samrabin/ctsm/python/empty.txt
Renaming existing pec_test_BVhWy to pec_test_BVhWy_orig
Renaming pec_test_BVhWy to pec_test_BVhWy_orig (this will take a few minutes)...

STDERR:
+ conda rename --force -n pec_test_BVhWy pec_test_BVhWy_orig
usage: conda [-h] [-v] [--no-plugins] [-V] COMMAND ...
conda: error: unrecognized arguments: --force

RETURN CODE: 2
```

`test_complete_py_env_create_mamba`:
```
FAIL: test_complete_py_env_create_mamba (test.test_sys_py_env_create.TestSysPyEnvCreate.test_complete_py_env_create_mamba)
A few calls of py_env_create to ensure it's working right with mamba instead of conda
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/glade/work/samrabin/ctsm/python/ctsm/test/test_sys_py_env_create.py", line 338, in test_complete_py_env_create_mamba
    self._create_empty_env(check=True, extra_args=["-m"])
    ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/glade/work/samrabin/ctsm/python/ctsm/test/test_sys_py_env_create.py", line 109, in _create_empty_env
    self.assertEqual(
    ~~~~~~~~~~~~~~~~^
        out.returncode, 0, "Unexpected failure of py_env_create call:\n" + out.stderr
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
AssertionError: 1 != 0 : Unexpected failure of py_env_create call:
+ mamba env create --yes --file /glade/work/samrabin/ctsm/python/empty.txt --name pec_test_zERMy
info     libmamba ****************** Backtrace Start ******************
debug    libmamba Loading configuration
trace    libmamba Compute configurable 'create_base'
trace    libmamba Compute configurable 'no_env'
trace    libmamba Compute configurable 'no_rc'
trace    libmamba Compute configurable 'rc_files'
trace    libmamba Compute configurable 'root_prefix'
trace    libmamba Get RC files configuration from locations up to HomeDir
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.mambarc'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.mamba/mambarc.d'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.mamba/mambarc'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.mamba/.mambarc'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.config/mamba/mambarc.d'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.config/mamba/mambarc'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.config/mamba/.mambarc'
trace    libmamba Configuration not found at '/glade/work/samrabin/conda-envs/ctsm_pylib/condarc.d'
trace    libmamba Configuration not found at '/glade/work/samrabin/conda-envs/ctsm_pylib/condarc'
trace    libmamba Configuration not found at '/glade/work/samrabin/conda-envs/ctsm_pylib/.condarc'
trace    libmamba Configuration found at '/glade/u/home/samrabin/.condarc'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.conda/condarc.d'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.conda/condarc'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.conda/.condarc'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.config/conda/condarc.d'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.config/conda/condarc'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.config/conda/.condarc'
trace    libmamba Configuration not found at '/glade/u/apps/opt/conda/.mambarc'
trace    libmamba Configuration not found at '/glade/u/apps/opt/conda/condarc.d'
trace    libmamba Configuration not found at '/glade/u/apps/opt/conda/condarc'
trace    libmamba Configuration found at '/glade/u/apps/opt/conda/.condarc'
trace    libmamba Configuration not found at '/var/lib/conda/.mambarc'
trace    libmamba Configuration not found at '/var/lib/conda/condarc.d/'
trace    libmamba Configuration not found at '/var/lib/conda/condarc'
trace    libmamba Configuration not found at '/var/lib/conda/.condarc'
trace    libmamba Configuration not found at '/etc/conda/.mambarc'
trace    libmamba Configuration not found at '/etc/conda/condarc.d/'
trace    libmamba Configuration not found at '/etc/conda/condarc'
trace    libmamba Configuration not found at '/etc/conda/.condarc'
trace    libmamba Update configurable 'no_env'
trace    libmamba Compute configurable 'envs_dirs'
trace    libmamba Compute configurable 'file_specs'
critical libmamba Got an empty file: /glade/work/samrabin/ctsm/python/empty.txt
info     libmamba ****************** Backtrace End ********************


Stdout:
cmd: /glade/work/samrabin/ctsm/py_env_create -n pec_test_zERMy -f /glade/work/samrabin/ctsm/python/empty.txt --yes -m
STDOUT:
Using mamba to install the python environment needed to run the CTSM python tools in the mamba environment: pec_test_zERMy
Using the file: /glade/work/samrabin/ctsm/python/empty.txt
Installing pec_test_zERMy; this can take a few minutes.

STDERR:
+ mamba env create --yes --file /glade/work/samrabin/ctsm/python/empty.txt --name pec_test_zERMy
info     libmamba ****************** Backtrace Start ******************
debug    libmamba Loading configuration
trace    libmamba Compute configurable 'create_base'
trace    libmamba Compute configurable 'no_env'
trace    libmamba Compute configurable 'no_rc'
trace    libmamba Compute configurable 'rc_files'
trace    libmamba Compute configurable 'root_prefix'
trace    libmamba Get RC files configuration from locations up to HomeDir
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.mambarc'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.mamba/mambarc.d'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.mamba/mambarc'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.mamba/.mambarc'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.config/mamba/mambarc.d'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.config/mamba/mambarc'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.config/mamba/.mambarc'
trace    libmamba Configuration not found at '/glade/work/samrabin/conda-envs/ctsm_pylib/condarc.d'
trace    libmamba Configuration not found at '/glade/work/samrabin/conda-envs/ctsm_pylib/condarc'
trace    libmamba Configuration not found at '/glade/work/samrabin/conda-envs/ctsm_pylib/.condarc'
trace    libmamba Configuration found at '/glade/u/home/samrabin/.condarc'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.conda/condarc.d'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.conda/condarc'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.conda/.condarc'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.config/conda/condarc.d'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.config/conda/condarc'
trace    libmamba Configuration not found at '/glade/u/home/samrabin/.config/conda/.condarc'
trace    libmamba Configuration not found at '/glade/u/apps/opt/conda/.mambarc'
trace    libmamba Configuration not found at '/glade/u/apps/opt/conda/condarc.d'
trace    libmamba Configuration not found at '/glade/u/apps/opt/conda/condarc'
trace    libmamba Configuration found at '/glade/u/apps/opt/conda/.condarc'
trace    libmamba Configuration not found at '/var/lib/conda/.mambarc'
trace    libmamba Configuration not found at '/var/lib/conda/condarc.d/'
trace    libmamba Configuration not found at '/var/lib/conda/condarc'
trace    libmamba Configuration not found at '/var/lib/conda/.condarc'
trace    libmamba Configuration not found at '/etc/conda/.mambarc'
trace    libmamba Configuration not found at '/etc/conda/condarc.d/'
trace    libmamba Configuration not found at '/etc/conda/condarc'
trace    libmamba Configuration not found at '/etc/conda/.condarc'
trace    libmamba Update configurable 'no_env'
trace    libmamba Compute configurable 'envs_dirs'
trace    libmamba Compute configurable 'file_specs'
critical libmamba Got an empty file: /glade/work/samrabin/ctsm/python/empty.txt
info     libmamba ****************** Backtrace End ********************

RETURN CODE: 1
```

--- a/py_env_create
+++ b/py_env_create
@@ -160,7 +160,9 @@ function is_version2_ge_version1 {
 # Handle an environment that already exists
 #
 function conda_env_exists {
-   ${condamamba} env list | grep -oE "/$1$" | wc -l
+   # Not using ${condamamba} because, at least as of mamba 2.3.1, it doesn't always show environment
+   # names. See https://github.com/mamba-org/mamba/issues/4045
+   conda env list | grep -oE "/$1$" | wc -l
 }
 function rename_existing_env {
    if [[ "$CONDA_DEFAULT_ENV" == *"$1" ]]; then

--- a/py_env_create
+++ b/py_env_create
@@ -143,6 +143,19 @@ if [ ! -f $condafile ]; then
    exit -1
 fi
 
+# Given two version number strings (unsigned integers separated by periods), echo "true" if the
+# second one is "greater than or equal to" the first. Otherwise, echo "false".
+# Based on https://unix.stackexchange.com/a/285928/208738
+function is_version2_ge_version1 {
+   requiredver=$1
+   currentver=$2
+   if [ "$(printf '%s\n' "$requiredver" "$currentver" | sort -V | head -n1)" = "$requiredver" ]; then
+      echo "true"
+   else
+      echo "false"
+   fi
+}
+
 #
 # Handle an environment that already exists
 #
@@ -155,7 +168,14 @@ function rename_existing_env {
       exit 1
    elif [[ $(conda_env_exists $2) -eq 0 ]]; then
       echo "Renaming $1 to $2 (this will take a few minutes)..."
-      yes_tmp=${yes/yes/force}
+      conda_version=$(conda --version | cut -d" " -f2)
+      if [[ "$(is_version2_ge_version1 25.3.0 ${conda_version})" == "true" ]]; then
+         yes_tmp=$yes
+      else
+         # Before conda 25.3.0, conda rename needed --force instead of --yes
+         # https://github.com/conda/conda/blob/408abc3a3826c80750883accb8896005e3ac97b2/CHANGELOG.md?plain=1#L291
+         yes_tmp=${yes/yes/force}
+      fi
       # Use conda instead of $condamamba because, at least as of mamba 1.5.9 / conda 24.7.1,
       # rename is not supported through mamba.
       if [[ "${quiet}" == "--quiet" ]]; then

--- a/python/ctsm/test/test_sys_py_env_create.py
+++ b/python/ctsm/test/test_sys_py_env_create.py
@@ -60,9 +60,11 @@ class TestSysPyEnvCreate(unittest.TestCase):
         self.py_env_create = os.path.join(path_to_ctsm_root(), "py_env_create")
         assert os.path.exists(self.py_env_create)
 
-        # Get path to testing condafile
+        # Get path to testing conda/mambafile:
+        # conda needs a completely empty file (# comments okay) as of 25.5.1, but mamba as of 2.3.1
+        # needs at least some YML structure.
         self.empty_condafile = os.path.join(path_to_ctsm_root(), "python", "empty.txt")
-        assert os.path.exists(self.empty_condafile)
+        self.empty_mambafile = os.path.join(path_to_ctsm_root(), "python", "empty.yml")
 
         # Set up other variables
         self.env_names = []
@@ -89,7 +91,12 @@ class TestSysPyEnvCreate(unittest.TestCase):
             self.env_names.append(get_unique_env_name(5))
 
         # Form and run command
-        cmd = [self.py_env_create, "-n", self.env_names[-1], "-f", self.empty_condafile, "--yes"]
+        if extra_args is not None and ("-m" in extra_args or "--mamba" in extra_args):
+            empty_file = self.empty_mambafile
+        else:
+            empty_file = self.empty_condafile
+        assert os.path.exists(empty_file)
+        cmd = [self.py_env_create, "-n", self.env_names[-1], "-f", empty_file, "--yes"]
         if extra_args:
             cmd += extra_args
         out = subprocess.run(cmd, capture_output=True, text=True, check=False)

--- a/python/ctsm/test/test_sys_py_env_create.py
+++ b/python/ctsm/test/test_sys_py_env_create.py
@@ -334,7 +334,8 @@ class TestSysPyEnvCreate(unittest.TestCase):
             raise e
         env_list = get_conda_envs()
         for env_name in self.env_names:
-            assert does_env_exist(env_name, env_list)
+            if not does_env_exist(env_name, env_list):
+                raise AssertionError(f"environment not found: {env_name}")
 
     def test_complete_py_env_create_mamba(self):
         """
@@ -365,7 +366,8 @@ class TestSysPyEnvCreate(unittest.TestCase):
             raise e
         env_list = get_conda_envs()
         for env_name in self.env_names:
-            assert does_env_exist(env_name, env_list)
+            if not does_env_exist(env_name, env_list):
+                raise AssertionError(f"environment not found: {env_name}")
 
 
 if __name__ == "__main__":

--- a/python/empty.yml
+++ b/python/empty.yml
@@ -1,4 +1,5 @@
-# An empty file for use in testing py_env_create
+# An empty file for use in testing py_env_create with mamba, which needs some YML structure
+# (as opposed to conda, which can use a truly empty file)
 #
 channels:
   - conda-forge

--- a/python/empty.yml
+++ b/python/empty.yml
@@ -1,0 +1,6 @@
+# An empty file for use in testing py_env_create
+#
+channels:
+  - conda-forge
+  - defaults
+dependencies:

